normalizer_lib.py

```diff
@@ -1,205 +1,224 @@
+        "Deletes freestanding punctuation."
+        delete_freestanding_punctuation = cdrewrite(
+            pynutil.delete(self.punctuation),
+            union("[BOS]", byte.SPACE),
+            union("[EOS]", byte.SPACE),
+            self.SIGMA_STAR)
+        return delete_freestanding_punctuation
+
+
+    def apply_fst_rules(self, string: str) -> str:
+        """Applies FST rewrite rules."""
+        return string if (string == "<STRING_REJECTED>") else (
+            string
+            @ self.language_specific_fixes()
+            @ self.detach_punctuation()
+            @ self.delete_freestanding_punctuation()
+            @ self.remove_extra_whitespace()
+            ).optimize().string()
+
+
+    def token_normalizer(self, string: str) -> str:
+        """Normalizes text by applying FST rewrite rules.
+
+        Args:
+            string: A line from a corpus.
+
+        Returns:
+            The normalized line from the corpus.
+        """
+        string = unicodedata.normalize("NFC", string.lower())
+        filtered_string = self.pass_only_valid_tokens(string)
+        return self.apply_fst_rules(filtered_string).strip()
 
-# Delete freestanding punctuation
-# e.g. "hi there ?" -> "hi there
 
-DELETE_FREESTANDING_PUNCTUATION = cdrewrite(
-    pynutil.delete(PUNCTUATION),
-    union("[BOS]", byte.SPACE),
-    union("[EOS]", byte.SPACE),
-    SIGMA_STAR)
+    def sentence_normalizer(self, string: str) -> str:
+        """Normalizes text by applying FST rewrite rules.
 
+        Args:
+            string: A line from a corpus.
 
-def apply_fst_rules(string: str) -> str:
-    """Applies FST rewrite rules."""
-    return string if (string == "<STRING_REJECTED>") else (
-        string
-        @ LANGUAGE_SPECIFIC_NORM
-        @ SEPARATE_PUNCTUATION
-        @ DELETE_FREESTANDING_PUNCTUATION
-        @ REMOVE_EXTRA_WHITESPACE
-        ).optimize().string().rstrip()
-
-
-def token_normalizer(string: str) -> str:
-    """Normalizes text by applying FST rewrite rules.
-
-    Args:
-        string: A line from a corpus.
-
-    Returns:
-        The normalized line from the corpus.
-    """
-    string = unicodedata.normalize("NFC", string.lower())
-    filtered_string = pass_only_valid_tokens(string)
-    return apply_fst_rules(filtered_string)
-
-
-def sentence_normalizer(string: str) -> str:
-    """Normalizes text by applying FST rewrite rules.
-
-    Args:
-        string: A line from a corpus.
-
-    Returns:
-        The normalized line from the corpus.
-    """
-    string = unicodedata.normalize("NFC", string.lower())
-    filtered_string = pass_only_valid_sentences(string)
-    return apply_fst_rules(filtered_string)
+        Returns:
+            The normalized line from the corpus.
+        """
+        string = unicodedata.normalize("NFC", string.lower())
+        filtered_string = self.pass_only_valid_sentences(string)
+        return self.apply_fst_rules(filtered_string).strip()
```

Comment:
Repeated logic could be in a helper method.

Reply:
I made a helper function `preprocess_string` to get rid of repeated logic.